### PR TITLE
[otel-tracing] Initial opentelemetry support. Rebase of 3701

### LINKED ISSUE #3451: Feature：add trace features
I hope to observe my mirror distribution process through opentracing, so I hope to add the Trace function. Do you have this requirement? Can you introduce it?

### PR DESCRIPTION
**Pull request by gotgelf**
This commit integrates the opentelemetry support functionality originally proposed in PR #3701 by @lengrongfu into the current main branch. Due to the age and fragility of the original PR, a direct rebase proved challenging. Therefore, the changes have been moved and adjusted to align with the latest codebase in main.

Key adjustments include:
- Aligning the tracing code with the latest changes in the main branch.
- Resolving conflicts.

All credit for the initial development of the tracing feature goes to @lengrongfu. This commit aims to build upon that foundation and bring this valuable feature into the current version of the project.

### Local Testing on macOS
**Jaeger Setup for Tracing**:
I ran a Jaeger instance locally:
```bash
docker run -d --name jaeger \
  -p 6831:6831/udp \
  -p 6832:6832/udp \
  -p 5775:5775/udp \
  -p 16686:16686 \
  -p 14268:14268 \
  -p 4317:4317 \
  jaegertracing/all-in-one:latest
```
After running my application configured to send trace data to Jaeger, I accessed the Jaeger UI at http://localhost:16686 to inspect the traces:

Screenshot 2023-12-07 at 17 16 16

Closes #3451

### CODE REVIEW

File: tracing/tracing.go [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+		),
+	)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create resource: %w", err)
+	}
+
+	// Configure OTLP trace exporter and set it up to connect to OpenTelemetry collector
+	// running on a local host.
+	ctrdTraceExporter, err := otlptracegrpc.New(ctx,
+		otlptracegrpc.WithInsecure(),
```
[the remark this answers is not in the document]
Comment (gotgelf) [reply tracing/tracing.go:47]: Hi, need, I just realised that. Initially I've seed all those comments as `outdated` and wrongly assumed that they were addressed. My fault.

**Comment by gotgelf** [comment]
> There are two distinct aspects to integrating OpenTelemetry into an application: instrumenting the code, and exporting traces. I recommend starting with a minimum viable product: a PR which [instruments one thing](https://pkg.go.dev/go.opentelemetry.io/contrib/instrumentation/github.com/gorilla/mux/otelmux), and [adds minimal support for exporting](https://pkg.go.dev/go.opentelemetry.io/contrib/exporters/autoexport).

I will start with that and create a first PR soon.

**Comment by corhere** [comment]
> I was wondering indeed if there should be any OTEL config options _at all_ in the configuration file.

My point exactly. This is the conversion I was trying to preempt. Unless you have any objections to autoexport, please hold your opinions on otel configuration until _after_ we get an MVP merged.

**Comment by milosgajdos** [comment]
This needs a rebase, assuming we still want to carry on with this PR.

**Comment by gotgelf** [comment]
> This needs a rebase, assuming we still want to carry on with this PR.

Hi @milosgajdos and @corhere,

Just to ensure we're all aligned, I'd like to reference a previous comment from @corhere:

> There are a lot of distinct separable changes all jammed together into that first commit. It is almost overwhelming to review. And it does more than the minimum required. Instead of trying to clean up this branch, how about we leave this PR as a resource to take parts from, and incrementally add OpenTelemetry instrumentation through multiple small, quick-to-review PRs?

Building on this approach:

- We've already initiated the first change with [PR #4188](https://github.com/distribution/distribution/pull/4188) (Otel tracing MVP).
- I'm currently working on the next PR(s), which aims to expand the instrumentation for storage driver(s).

Once these are merged, we can consider closing the current PR along with all other outdated ones related to OTEL. The primary objective for keeping this PR open, for now, is to track all desired changes and ensure they are systematically covered in smaller, separate PRs.

Could you please confirm if this aligns with your expectations, or if there's anything else I might have overlooked?

**Comment by milosgajdos** [comment]
That sounds good to me. Yeah we need to close them out. Maybe we should close this straight away and the follow up with separate ones. Feel free to do so.

**Comment by gotgelf** [comment]
There are a lot of distinct separable changes all jammed together into that first commit. It is almost overwhelming to review. And it does more than the minimum required. Instead of trying to clean up this branch, how about we leave this PR as a resource to take parts from, and incrementally add OpenTelemetry instrumentation through multiple small, quick-to-review PRs

**Pull request closed without merging**